File: educational-vms/string_rewriting/software_tools_m4/gettok.r
include /usr/style/io/globdefs
include defdefs
# gettok _ get alphanumeric string or single non\(hyalpha for define
	character function gettok(token, toksiz)
	character ngetc, type
	integer i, toksiz
	character token(toksiz)

	for (i = 1; i < toksiz; i = i + 1) {
		gettok = type(ngetc(token(i)))
		if (gettok != LETTER & gettok != DIGIT)
			break
		}
	if (i >= toksiz)
		call error("token too long.")
	if (i > 1) {			# some alpha was seen
		call putbak(token(i))
		i = i - 1
		gettok = ALPHA
		}
	# else single character token
	token(i+1) = EOS
	return
	end

File: educational-vms/string_rewriting/software_tools_m4/gettok.c
#include "gettok.h"

#include "globdefs.h"
#include "defdefs.h"
#include <ctype.h>

// get alphanumeric string or single non-alpha for define
char gettok(char* token, int toksiz) {
  int i;
  char return_value;
  char c;

  for (i= 1; i < toksiz; ++i) {
    c= ngetc(token[i]);
    if (isalpha(c)) {
      return_value= LETTER;
    } else if (isdigit(c)) {
      return_value= DIGIT;
    } else {
      return_value= OTHER;
    }
    if (return_value != LETTER && return_value != DIGIT) {
      break;
    }
  }
  if (i >= toksiz) {
    error("token too long.");
  }
  if (i > 1) { // some alphanum was seen
    putbak(token[i]);
    --i;
    return_value= ALPHA;
  }
  // single character token
  token[i+1]= EOS;
  return return_value;
}
    
